### PR TITLE
Clean up native row accessor from the same thread that created them

### LINKED ISSUE #1311: Destruction of mutex in use  Fatal signal 6 (SIGABRT) (code=-6)
GC during Iterating a huge RealmResults might cause a crash.

Sample project to reproduce this issue can be found https://drive.google.com/open?id=0B2It_kNwIwTBUHV1WC01V0w3LWs

To reproduce it, click `populate database` and then click `process` multiple times.

Crash logs here.

F/libc    (17612): Fatal signal 6 (SIGABRT) at 0x000044cc (code=-6), thread 17612 (.realmcrushtest)
I/DEBUG   (  330): *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
I/DEBUG   (  330): UUID: a666c02e-3d9a-4a5c-a11c-075aced1faa8
I/DEBUG   (  330): Build fingerprint: 'Sony/C6833/C6833:4.4.4/14.4.A.0.108/k___jQ:user/release-keys'
I/DEBUG   (  330): Revision: '0'
I/DEBUG   (  330): pid: 17612, tid: 17612, name: .realmcrushtest  >>> com.arturkinkel.realmcrushtest <<<
I/DEBUG   (  330): signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
I/DEBUG   (  330):     r0 00000000  r1 000044cc  r2 00000006  r3 00000000
I/DEBUG   (  330):     r4 00000006  r5 00000000  r6 000044cc  r7 0000010c
I/DEBUG   (  330):     r8 000000a6  r9 793e6e1c  sl 793e6e24  fp beef139c
I/DEBUG   (  330):     ip 7909a64e  sp beef1178  lr 4012616d  pc 40135118  cpsr 000f0010
I/DEBUG   (  330):     d0  742065657320756f  d1  6f72726520736968
I/DEBUG   (  330):     d2  7361656c70202c72  d3  7420646e65732065
I/DEBUG   (  330):     d4  0000000200008000  d5  0400000000000000
I/DEBUG   (  330):     d6  3e4ccccd00b13000  d7  42c800004c232614
I/DEBUG   (  330):     d8  43e928b800000800  d9  3f40000000000600
I/DEBUG   (  330):     d10 0000000000000600  d11 0000000000000000
I/DEBUG   (  330):     d12 0000000000000000  d13 0000000000000000
I/DEBUG   (  330):     d14 0000000000000000  d15 0000000000000000
I/DEBUG   (  330):     d16 20676f6c20736968  d17 40706c6568206f74
I/DEBUG   (  330):     d18 736544203a363631  d19 6e6f697463757274
I/DEBUG   (  330):     d20 6574756d20666f20  d21 657375206e692078
I/DEBUG   (  330):     d22 4154524f504d490a  d23 79206669203a544e
I/DEBUG   (  330):     d24 0018001700150016  d25 0019001a001a0019
I/DEBUG   (  330):     d26 0000000000000000  d27 0000000000000000
I/DEBUG   (  330):     d28 3ff0000000000000  d29 0000000000000000
I/DEBUG   (  330):     d30 001a001a001a001a  d31 0000000000000000
I/DEBUG   (  330):     scr 20000012
I/DEBUG   (  330): 
I/DEBUG   (  330): backtrace:
I/DEBUG   (  330):     #00  pc 00022118  /system/lib/libc.so (tgkill+12)
I/DEBUG   (  330):     #01  pc 00013169  /system/lib/libc.so (pthread_kill+48)
I/DEBUG   (  330):     #02  pc 0001337d  /system/lib/libc.so (raise+10)
I/DEBUG   (  330):     #03  pc 000120b3  /system/lib/libc.so
I/DEBUG   (  330):     #04  pc 000219cc  /system/lib/libc.so (abort+4)
I/DEBUG   (  330):     #05  pc 00042e01  /data/app-lib/com.arturkinkel.realmcrushtest-1/librealm-jni.so
I/DEBUG   (  330):     #06  pc 0004373f  /data/app-lib/com.arturkinkel.realmcrushtest-1/librealm-jni.so
I/DEBUG   (  330): 
I/DEBUG   (  330): stack:
I/DEBUG   (  330):          beef1138  c3c2c1c0  
I/DEBUG   (  330):          beef113c  000000a6  
I/DEBUG   (  330):          beef1140  00000020  
I/DEBUG   (  330):          beef1144  beef1218  [stack]
I/DEBUG   (  330):          beef1148  00000086  
I/DEBUG   (  330):          beef114c  790af7bc  
I/DEBUG   (  330):          beef1150  00000093  
I/DEBUG   (  330):          beef1154  00000168  
I/DEBUG   (  330):          beef1158  7955e170  
I/DEBUG   (  330):          beef115c  40123ff3  /system/lib/libc.so (dlmalloc+4254)
I/DEBUG   (  330):          beef1160  00000086  
I/DEBUG   (  330):          beef1164  0080000c  
I/DEBUG   (  330):          beef1168  40122f55  /system/lib/libc.so (dlmalloc)
I/DEBUG   (  330):          beef116c  00000086  
I/DEBUG   (  330):          beef1170  4015d1d8  /system/lib/libc.so
I/DEBUG   (  330):          beef1174  727833f8  
I/DEBUG   (  330):     #00  beef1178  00000006  
I/DEBUG   (  330):          beef117c  00000000  
I/DEBUG   (  330):          beef1180  000044cc  
I/DEBUG   (  330):          beef1184  7909a64e  /data/app-lib/com.arturkinkel.realmcrushtest-1/librealm-jni.so
I/DEBUG   (  330):          beef1188  7909a64e  /data/app-lib/com.arturkinkel.realmcrushtest-1/librealm-jni.so
I/DEBUG   (  330):          beef118c  4012616d  /system/lib/libc.so (pthread_kill+52)
I/DEBUG   (  330):     #01  beef1190  00000006  
I/DEBUG   (  330):          beef1194  00000000  
I/DEBUG   (  330):          beef1198  7909aa92  /data/app-lib/com.arturkinkel.realmcrushtest-1/librealm-jni.so
I/DEBUG   (  330):          beef119c  40126381  /system/lib/libc.so (raise+14)
I/DEBUG   (  330):     #02  beef11a0  beef11ac  [stack]
I/DEBUG   (  330):          beef11a4  401250b7  /system/lib/libc.so
I/DEBUG   (  330): 
I/DEBUG   (  330): memory near r1:
I/DEBUG   (  330):     000044ac ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     000044bc ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     000044cc ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     000044dc ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     000044ec ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     000044fc ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000450c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000451c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000452c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000453c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000454c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000455c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000456c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000457c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000458c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000459c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330): 
I/DEBUG   (  330): memory near r6:
I/DEBUG   (  330):     000044ac ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     000044bc ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     000044cc ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     000044dc ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     000044ec ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     000044fc ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000450c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000451c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000452c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000453c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000454c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000455c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000456c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000457c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000458c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330):     0000459c ffffffff ffffffff ffffffff ffffffff  
I/DEBUG   (  330): 
I/DEBUG   (  330): memory near r9:
I/DEBUG   (  330):     793e6dfc 00000000 00000001 0025d100 00300000  
I/DEBUG   (  330):     793e6e0c 793e6de8 00000020 00000039 00000000  
I/DEBUG   (  330):     793e6e1c 77d31000 00000458 790cc000 00000458  
I/DEBUG   (  330):     793e6e2c 00000001 77c7db4c 7163db1c 00000000  
I/DEBUG   (  330):     793e6e3c 00000001 77d31034 00000000 77d31038  
I/DEBUG   (  330):     793e6e4c 00000000 77d3103c 00000000 77d31040  
I/DEBUG   (  330):     793e6e5c 00000000 79420090 0000001b 00000000  
I/DEBUG   (  330):     793e6e6c 0000ffd5 793e6e80 77cd8da0 00000018  
I/DEBUG   (  330):     793e6e7c 0000001b 00000000 0000ffd6 793e0a58  
I/DEBUG   (  330):     793e6e8c 793e6e68 00000018 0000001b 6d7cafc0  
I/DEBUG   (  330):     793e6e9c 00000044 00000000 00000000 00000001  
I/DEBUG   (  330):     793e6eac 0000001b 00000000 00010759 793e6ec8  
I/DEBUG   (  330):     793e6ebc 79418768 77cd84e0 0000001b 00000000  
I/DEBUG   (  330):     793e6ecc 0001075a 793e6ee0 793e6eb0 00000030  
I/DEBUG   (  330):     793e6edc 0000001b 00000000 0001075b 793e6ef8  
I/DEBUG   (  330):     793e6eec 793e6ec8 00000048 0000001b 00000000  
I/DEBUG   (  330): 
I/DEBUG   (  330): memory near sl:
I/DEBUG   (  330):     793e6e04 0025d100 00300000 793e6de8 00000020  
I/DEBUG   (  330):     793e6e14 00000039 00000000 77d31000 00000458  
I/DEBUG   (  330):     793e6e24 790cc000 00000458 00000001 77c7db4c  
I/DEBUG   (  330):     793e6e34 7163db1c 00000000 00000001 77d31034  
I/DEBUG   (  330):     793e6e44 00000000 77d31038 00000000 77d3103c  
I/DEBUG   (  330):     793e6e54 00000000 77d31040 00000000 79420090  
I/DEBUG   (  330):     793e6e64 0000001b 00000000 0000ffd5 793e6e80  
I/DEBUG   (  330):     793e6e74 77cd8da0 00000018 0000001b 00000000  
I/DEBUG   (  330):     793e6e84 0000ffd6 793e0a58 793e6e68 00000018  
I/DEBUG   (  330):     793e6e94 0000001b 6d7cafc0 00000044 00000000  
I/DEBUG   (  330):     793e6ea4 00000000 00000001 0000001b 00000000  
I/DEBUG   (  330):     793e6eb4 00010759 793e6ec8 79418768 77cd84e0  
I/DEBUG   (  330):     793e6ec4 0000001b 00000000 0001075a 793e6ee0  
I/DEBUG   (  330):     793e6ed4 793e6eb0 00000030 0000001b 00000000  
I/DEBUG   (  330):     793e6ee4 0001075b 793e6ef8 793e6ec8 00000048  
I/DEBUG   (  330):     793e6ef4 0000001b 00000000 0001075c 793e6f10  
I/DEBUG   (  330): 
I/DEBUG   (  330): memory near fp:
I/DEBUG   (  330):     beef137c 4191a4e8 beef139c 41597350 4164acb8  
I/DEBUG   (  330):     beef138c 00000001 00000008 43a507a0 00000025  
I/DEBUG   (  330):     beef139c 415c7fd3 4164acb8 75c5ba7a 79065095  
I/DEBUG   (  330):     beef13ac 4191a4e8 4163ae18 4164ac78 73300019  
I/DEBUG   (  330):     beef13bc 00000000 ffffffff 40161394 4191a3e0  
I/DEBUG   (  330):     beef13cc 415ed141 41634c38 41634c38 40161394  
I/DEBUG   (  330):     beef13dc 00000010 4163ae18 40161394 00000010  
I/DEBUG   (  330):     beef13ec 415a398c 00000000 00000000 00000000  
I/DEBUG   (  330):     beef13fc 00000000 00000000 00000000 00000000  
I/DEBUG   (  330):     beef140c 00000000 00000000 00000000 00000000  
I/DEBUG   (  330):     beef141c 00000000 00000000 00000000 00000000  
I/DEBUG   (  330):     beef142c 00000000 00000000 00000000 00000000  
I/DEBUG   (  330):     beef143c 00000000 00000000 00000000 00000000  
I/DEBUG   (  330):     beef144c 00000000 00000000 8e8c99ae 00000001  
I/DEBUG   (  330):     beef145c 41634c38 4191be90 00000000 00000010  
I/DEBUG   (  330):     beef146c 00000000 4164ac78 415e2fcf 4191a4d8  
I/DEBUG   (  330): 
I/DEBUG   (  330): memory near ip:
I/DEBUG   (  330):     7909a62c 41455200 52004d4c 72756365 65766973  
I/DEBUG   (  330):     7909a63c 636f6c20 676e696b 20666f20 6574756d  
I/DEBUG   (  330):     7909a64c 74750078 742f6c69 61657268 70632e64  
I/DEBUG   (  330):     7909a65c 74700070 61657268 756d5f64 5f786574  
I/DEBUG   (  330):     7909a66c 6b636f6c 66202928 656c6961 725b0064  
I/DEBUG   (  330):     7909a67c 6d6c6165 726f632d 2e302d65 352e3938  
I/DEBUG   (  330):     7909a68c 7341205d 74726573 206e6f69 6c696166  
I/DEBUG   (  330):     7909a69c 203a6465 3d3d2072 20003020 72726500  
I/DEBUG   (  330):     7909a6ac 635f726f 2065646f 3b00203d 65720020  
I/DEBUG   (  330):     7909a6bc 646f6374 203d2065 72745300 44676e69  
I/DEBUG   (  330):     7909a6cc 2e617461 657a6973 00203d20 69727453  
I/DEBUG   (  330):     7909a6dc 6144676e 642e6174 20617461 5300203d  
I/DEBUG   (  330):     7909a6ec 6e697274 74614467 73612061 78656820  
I/DEBUG   (  330):     7909a6fc 00203d20 00783020 625f6e69 6e696765  
I/DEBUG   (  330):     7909a70c 00203d20 655f6e69 3d20646e 756f0020  
I/DEBUG   (  330):     7909a71c 75635f74 3d207272 756f0020 6e655f74  
I/DEBUG   (  330): 
I/DEBUG   (  330): memory near sp:
I/DEBUG   (  330):     beef1158 7955e170 40123ff3 00000086 0080000c  
I/DEBUG   (  330):     beef1168 40122f55 00000086 4015d1d8 727833f8  
I/DEBUG   (  330):     beef1178 00000006 00000000 000044cc 7909a64e  
I/DEBUG   (  330):     beef1188 7909a64e 4012616d 00000006 00000000  
I/DEBUG   (  330):     beef1198 7909aa92 40126381 beef11ac 401250b7  
I/DEBUG   (  330):     beef11a8 000000a6 ffffffdf 4015d000 beef11e0  
I/DEBUG   (  330):     beef11b8 beef11e0 beef11ec 7909aa92 beef11e4  
I/DEBUG   (  330):     beef11c8 beef11ec 401349d0 7955e178 7901ee05  
I/DEBUG   (  330):     beef11d8 7955e178 7901f743 7955e17c 790a823c  
I/DEBUG   (  330):     beef11e8 00000000 790a8250 790a80f0 77d08b1c  
I/DEBUG   (  330):     beef11f8 77d08b1c 77d08b1d 77d08b1c 77d08b95  
I/DEBUG   (  330):     beef1208 77d08d1c 790ae984 00000018 77d08b1c  
I/DEBUG   (  330):     beef1218 790a8264 00000006 00000000 00001002  
I/DEBUG   (  330):     beef1228 00000000 00000000 00000000 00000000  
I/DEBUG   (  330):     beef1238 00000000 00000000 00000000 00000000  
I/DEBUG   (  330):     beef1248 00000000 00000000 00000000 00000000  
I/DEBUG   (  330): 
I/DEBUG   (  330): code around pc:
I/DEBUG   (  330):     401350f8 e8bd00f0 e3700a01 912fff1e e2600000  
I/DEBUG   (  330):     40135108 ea006e2e e92d50f0 e3a07f43 ef000000  
I/DEBUG   (  330):     40135118 e8bd50f0 e3700a01 912fff1e e2600000  
I/DEBUG   (  330):     40135128 ea006e26 e92d50f0 e3a070ee ef000000  
I/DEBUG   (  330):     40135138 e8bd50f0 e3700a01 912fff1e e2600000  
I/DEBUG   (  330):     40135148 ea006e1e f200429a bf0080b9 f040f891  
I/DEBUG   (  330):     40135158 4001e92d f2c02a04 2a1080a5 8093f2c0  
I/DEBUG   (  330):     40135168 f2c02a20 2a408088 ea4fdb7f f1bc1c92  
I/DEBUG   (  330):     40135178 dd6c0f0a 0600e92d 0f40f1bc f500dd4a  
I/DEBUG   (  330):     40135188 f5016e80 ebae7920 ea4f0e09 ea4f5e4e  
I/DEBUG   (  330):     40135198 f50e5e5e ebbc7e20 dd3b1f9e f04fbfc4  
I/DEBUG   (  330):     401351a8 ebd9090a dd35199e 0a0eeb01 0a3ff02a  
I/DEBUG   (  330):     401351b8 1c9eebac bfd245e1 0c09ebac f04f46e1  
I/DEBUG   (  330):     401351c8 f8910c00 f891f240 f921f280 f921028d  
I/DEBUG   (  330):     401351d8 f8da428d f1b93000 f9000901 f900028d  
I/DEBUG   (  330):     401351e8 f10a428d d1ee0a40 0f00f1bc f5bcd02b  
I/DEBUG   (  330): 
I/DEBUG   (  330): code around lr:
I/DEBUG   (  330):     4012614c 447b4b13 42b3e010 6a1ed10e 44784811  
I/DEBUG   (  330):     4012615c ec84f7fb ea0ef00d 46224631 efd0f00e  
I/DEBUG   (  330):     4012616c d00a3001 e00b2400 2b00681b 480ad1eb  
I/DEBUG   (  330):     4012617c 44782403 ec72f7fb f001e002 6804fa23  
I/DEBUG   (  330):     4012618c fa20f001 46206005 bf00bd70 0003b266  
I/DEBUG   (  330):     4012619c 0003b25a 0003b252 0003b22e bf7ef7ff  
I/DEBUG   (  330):     401261ac 4a3e4b3d e92d447b b08b43f0 4606589c  
I/DEBUG   (  330):     401261bc 6823460d 930946a1 fa04f001 8000f8d0  
I/DEBUG   (  330):     401261cc d0482d00 f0104628 280ff91d d8444604  
I/DEBUG   (  330):     401261dc ffe4f7ff d1064286 4629200f e8d8f00d  
I/DEBUG   (  330):     401261ec d03c2800 482ee02e f7fb4478 482debb6  
I/DEBUG   (  330):     401261fc e0154478 d11342b0 482b6a06 f7fb4478  
I/DEBUG   (  330):     4012620c 4a2aec2e 46332120 a801447a f9eef013  
I/DEBUG   (  330):     4012621c a8012101 fd0cf01a 46061c42 e011d104  
I/DEBUG   (  330):     4012622c 28006800 e02cd1e6 46294630 f00d4622  
I/DEBUG   (  330):     4012623c 1c43e914 d11e4607 f9c4f001 29046801

### PR DESCRIPTION
**Pull request by emanuelez**
Fixes #1311

### CODE REVIEW

**Comment by emanuelez** [comment]
@realm/java @kspangsege

**Comment by kspangsege** [comment]
:+1:

File: realm/src/main/java/io/realm/Realm.java [hunk 1]
```diff
@@ -573,12 +572,6 @@ private static Realm create(RealmConfiguration configuration) {
     }
 
     private static synchronized Realm createAndValidate(RealmConfiguration configuration, boolean validateSchema, boolean autoRefresh) {
-        // Start the finalizer thread if needed
-        if (!isFinalizerStarted) {
-            executorService.submit(new FinalizerRunnable());
-            isFinalizerStarted = true;
-        }
-
         // Check if a cached instance already exists for this thread
```
Comment (beeender) [inline realm/src/main/java/io/realm/Realm.java]: isFinalizerStarted can be removed.

File: realm/src/main/java/io/realm/internal/Context.java [hunk 2]
```diff
@@ -52,6 +57,17 @@ public void executeDelayedDisposal() {
...
+            cleanRows();
+        }
+    }
+
+    public void cleanRows() {
+        NativeObjectReference reference = (NativeObjectReference) referenceQueue.poll();
+        while (reference != null) {
+            UncheckedRow.nativeClose(reference.nativePointer);
+            rowReferences.remove(reference);
+            reference = (NativeObjectReference) referenceQueue.poll();
```
Comment (beeender) [inline realm/src/main/java/io/realm/internal/Context.java]: if `Context.isFinalized` is usefull, we probably need to check it in this function as well.

Comment (emanuelez) [inline realm/src/main/java/io/realm/internal/Context.java]: No, it is only used when adding to the list, not when cleaning them up.

**Comment by emanuelez** [comment]
retest this please

**Comment by beeender** [comment]
:+1: So bad we still cannot get rid of context :cry:

**Comment by beeender** [comment]
retest this please

**Comment by beeender** [comment]
:+1:

**Pull request merged**